Fix #2284: Schnorr signatures misusage leaks private keys

Geod24 opened the pull request:
Based on #2552

Only the last commit belongs to this PR.
```
This effectively changes the formula used for signing blocks, turning it upside-down:
Instead of offsetting the 'R' by the pre-image as a point,
we commit to the pre-image being the 's' of the signature,
and provide 'R' such as 'S == c * R + X'.
In order to find the adequate 'R' value, one has to know both 'x' and 'r',
as one may only derive CR from the known values, and CR cannot be divided by 'c'.
In the process of this change, the no-longer-used 'getCommitmentNonce' was removed.
```

Geod24 committed:
Fix #2284: Schnorr signatures misusage leaks private keys

This effectively changes the formula used for signing blocks, turning it upside-down:
Instead of offsetting the 'R' by the pre-image as a point,
we commit to the pre-image being the 's' of the signature,
and provide 'R' such as 'S == c * R + X'.
In order to find the adequate 'R' value, one has to know both 'x' and 'r',
as one may only derive CR from the known values, and CR cannot be divided by 'c'.
In the process of this change, the no-longer-used 'getCommitmentNonce' was removed.

hewison-chris reviewed (approved):
Excellent :-)

hewison-chris reviewed (changes requested):
Block IO unit tests also need to be updated.

Geod24 committed:
Trivial: Remove redundant imports

This is already imported at the top level (module scope).

Geod24 committed:
Trivial: Shorted creeateBlockSignature's names

Simply use 'sign' in the BlockHeader, and a slightly more verbose name in the Nominator.
This is motivated by the desire to introduce a 'verify' function which would mirror it.

Geod24 committed:
BlockHeader: Add verify methods, mirroring sign

hewison-chris commented on source/agora/consensus/data/Block.d:
```diff
@@ -892,8 +903,8 @@ unittest
+    assert(block2.header.verify(genesis_validator_keys[0].address, block1.header.preimages[0], sig2.R));
 
     // Calculate the private key by subtraction
     // `s = r + (v * c)`
```
This will need to be updated to show it still works.

hewison-chris reviewed (approved):
LGTM just a small update to show revealing private key when signing two blocks at same height.

hewison-chris commented:
<img width="962" alt="Screen Shot 2021-10-07 at 10 13 09" src="https://user-images.githubusercontent.com/11815575/136305160-f7f2b88d-dc9f-4791-8d55-dcb32bd73600.png">
